docs/programming-guide.md

```diff
@@ -728,6 +728,69 @@ def doStuff(self, rdd):
+
+However, in `cluster` mode, what happens is more complicated, and the above code will not work correctly. To execute jobs Spark breaks up the processing of RDD operations into tasks - each of which is operated on by an executor. Prior to execution, Spark computes the **closure**. The closure is those variables and methods which must be visible for the executor to perform its computations on the RDD (in this case `foreach()`). This closure is serialized and sent to each executor. In `local` mode, there is only the one executors so everything shares the same closure. In `remote` mode however, this is not the case and the executors running on seperate worker nodes each have their own copy of the closure.
+
+The problem here is that the variables within the closure sent to each executor are now copies and thus, when **counter** is referenced within the `foreach` function, it's no longer the **counter** on the driver node. There is still a **counter** in the memory of the driver node but this is no longer visible to the executors! The executors only sees the copy from the serialized closure. Thus, the final value of **counter** will still be zero since all operations on **counter** were referencing the value within the serialized closure.  
+
+The one exception to this is when the variable being modified is an [`Accumulator`](#AccumLink). Accumulators in Spark are used specifically to provide a mechanism for safely updating a variable when execution is split up across worker nodes in a cluster. The Accumulators section of this guide discusses these in more detail.  
+
+In general, closures - constructs like loops or locally defined methods, should not be used to mutate some global state. Spark does not define or guarantee the behavior of mutations to objects referenced from outside of closures. Some code that does this may work in local mode, but that's just by accident and such code will not behave as expected in distributed mode. Use an accumulator instead if some global aggregation is needed.
+
+#### Printing elements of an RDD 
+Another common idiom is attempting to print out the elements of an RDD using `rdd.foreach(println)` or `rdd.map(println)`. Intuitively, it seems that this should work. But again, consider that in `cluster` mode, the output to `stdout` being called by the executors is now writing to the executor's `stdout` instead, not the one on the driver! Consequently, Spark will start writing to `stdout` on the worker node, potentially filling up `/tmp` storage rapdily. To avoid this, one can use the `collect()` method to first bring the RDD to the driver node thus: `rdd.collect().foreach(println)`. This can cause the driver to run out of memory, though, because collect() fetches the entire RDD to a single machine; a safer approach is to use the `take()` method to only get a few elements of the RDD: `rdd.take(100).foreach(println)`.
```

Comment:
Same hair splitting as in the last review - I don't know if it's intuitive or not. It's behavior that I would expect, but that will surprise some.

I don't think the problem here is filling up `/tmp` right? I mean, that's fairly orthogonal. I would suggest removing this as it might distract from the issue. Neither of these is a problem per se, but only one operation probably does what you want.